File: PythonSpark/JobManagment/Job.py
import argparse
from abc import ABC, abstractmethod

from pyspark.sql import SparkSession, DataFrame

from Utils.RichArgumentParser import RichArgumentParser


class Job(ABC):
    def __init__(self, arguments: argparse.Namespace, spark: SparkSession):
        self.arguments = arguments
        self.spark = spark

    @abstractmethod
    def load(self) -> dict:
        pass

    @abstractmethod
    def process(self, data: dict) -> DataFrame:
        pass

    @abstractmethod
    def write(self, data: DataFrame) -> None:
        pass

    @staticmethod
    def parser() -> RichArgumentParser:
        return RichArgumentParser("Job")

File: ScalaSpark/src/scala/JobManagment/Job.scala
package JobManagment

import de.halcony.argparse.Parser
import org.apache.spark.sql.{DataFrame, SparkSession}

abstract class Job(arguments: Map[String, String], spark: SparkSession) {
  def load(): Map[String, DataFrame]

  def process(data: Map[String, DataFrame]): DataFrame

  def write(data: DataFrame): Unit
}

object Job extends JobCompanion {
  override def parser(): Parser = Parser("Job")
}